java/ql/src/semmle/code/java/security/UnsafeDeserialization.qll

```diff
@@ -109,6 +116,150 @@ class SafeKryo extends DataFlow2::Configuration {
   }
 }
 
+/**
+ * A class that overrides the `android.os.Parcelable.Creator.createFromParcel` method
+ * and has the Gson call.
+ */
+class CreateFromParcelMethod extends Method {
+  CreateFromParcelMethod() {
+    this.hasName("createFromParcel") and
+    this.getEnclosingCallable().getDeclaringType().getASupertype*() instanceof TypeParcelable
+  }
+}
+
+/** Unsafe Gson configuration with dynamic type. */
+class UnsafeGsonConfig extends TaintTracking2::Configuration {
+  UnsafeGsonConfig() { this = "UnsafeDeserialization::UnsafeGsonConfig" }
+
+  override predicate isSource(DataFlow::Node src) { src instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof GsonDeserializeMethod and
+      sink.asExpr() = ma.getArgument(1) // The class type argument
+    )
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(ReflectiveClassIdentifierMethodAccess ma |
+      node1.asExpr() = ma.getArgument(0) and
+      node2.asExpr() = ma
+    )
+    or
+    exists(CreateFromParcelMethod m, Variable v |
+      m.getEnclosingCallable().getDeclaringType() = v.getType() and
+      node1.asExpr() = v.getAnAssignedValue() and
+      node2.asExpr() = m.getAParameter().getAnAccess()
+    )
+  }
+}
+
+/** Source of creating a `gson` instance using the builder. */
+class CreateGsonSource extends DataFlow::ExprNode {
+  CreateGsonSource() {
+    exists(MethodAccess ma | ma = this.asExpr() |
+      ma.getMethod() instanceof CreateGsonMethod and
+      exists(MethodAccess rma |
+        rma.getMethod() instanceof RegsterClassTypeMethod // register allowed class types using the builder
```

Comment:
These methods only allow registering custom adapters, however they do not restrict deserialization to these types.
So if the deserialized type is user-controlled, then this most likely does not make the deserialization safe; unless a `TypeAdapterFactory` or similar is registered and it is filtering types, but that is probably rather uncommon.

Reply:
Yes, I also think it's rather uncommon to allow dynamic types while configuring specific types in a <code>TypeAdapterFactory</code>. And <code>TypeAdapterFactory</code> does not make the deserialization safe here when the deserialized type is user-controlled so I will remove it in a separate commit.

Reply:
Thanks @Marcono1234 again for reviewing this query. I've committed the change.